internal/graph/check.go

```diff
@@ -38,6 +39,12 @@ type ResolveCheckRequest struct {
 	Context              *structpb.Struct
 	ResolutionMetadata   *ResolutionMetadata
 	VisitedPaths         map[string]struct{}
+	// This is different from ResolutionMetadata dispatch counter in that this counter needs to be
+	// incremented before dispatches are processed (where the ResolutionMetadata dispatch counter
+	// are incremented after the dispatches are processed). As this counter needs to be updated
+	// synchronously are multiple concurrent go-routine, it needs to use shared counter within
+	// the same request.
+	DispatchCounter *atomic.Uint32
```

Comment:
Updating the dispatch counter to avoid using atomic is the next item we'll look into.  At that time, we should have a single counter instead of two counters.

Reply:
I want to defer this part of work into separate PR to avoid expanding this PR's scope.  The work for this is already ongoing.

Reply:
But it wouldn't expand this PR scope. Instead of doing `currentNumDispatch := req.DispatchCounter.Add(1)` you would do 

```go
resp, err := r.delegate.ResolveCheck(ctx, req)
if err != nil {
    // ...
}
currentNumDispatch := resp.GetResolutionMetadata().GetDispatchCount()
```

would that not work?

Reply:
Not quite: We need to increment the counter when we decide whether we want to do dispatch throttling or not.  For example, imagine you have two simultaneous dispatches.

Normally, you would have expect one dispatch will have dispatch count of X and the other to be X + 1.  However, in the sample code above, it is possible that you will observe both to be X+1 because currentNumDispatch is measured after it is incremented. This is not necessary a bad thing - but this warrants consideration of where we want to really increment that counter.  

This is why I propose to defer this to a subsequent PR

Reply:
Separate counters are necessary to complete the two tasks at hand – throttling and reporting. Count on request is to signal to determine wether to throttle as the dispatches expand recursively. Those dispatches could fan-out into hundreds more. That count is necessary at the time of dispatch to make that decision. Response on the other hand is returning dispatch count for reporting/metrics. By the time the response happens, its too late for throttling to occur. 

In short, it makes sense to me to keep these separate. I'm not sure how you'd combine.

Reply:
I understand now. This is a global counter. I have a proposal in https://github.com/openfga/openfga/pull/1453/files so that we put this counter within the request metadata

Reply:
>Separate counters are necessary to complete the two tasks at hand – throttling and reporting. Count on request is to signal to determine wether to throttle as the dispatches expand recursively. Those dispatches could fan-out into hundreds more. That count is necessary at the time of dispatch to make that decision. Response on the other hand is returning dispatch count for reporting/metrics. By the time the response happens, its too late for throttling to occur.

Exactly.  Not that it cannot be done - but let's defer this to a different PR so that we can focus on the dispatch throttling piece.

Reply:
If you're actively in `DispatchThrottlingCheckResolver#ResolveCheck`, then you'e already dispatched the subproblem. You've gone through 2 layers of the onion already `CycleDetectionCheckResolver` and `DispatchThrottlingCheckResolver`. In the future this could even mean that you've distributed the problem to another remote server. Consequently, incrementing the dispatch count should happen as you are dispatching it, not after you've dispatched.